sqlstats/sqlstats.go

```diff
@@ -0,0 +1,128 @@
+// Package sqlstats implements an SQLite Tracer that collects query stats.
+package sqlstats
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/tailscale/sqlite"
+)
+
+// Tracer implements sqlite.Tracer and collects query stats.
+//
+// To use, pass the tracer object to sqlite.Connector, then start a debug
+// web server with http.HandlerFunc(sqlTracer.Handle).
+type Tracer struct {
+	mu      sync.RWMutex
+	queries map[string]*queryStats // query -> stats
+}
+
+type queryStats struct {
+	query string
+
+	// When inside the queries map all fields must be accessed as atomics.
+	count    int64
+	errors   int64
+	duration int64 // time.Duration
+	// TODO lastErr atomic.Value
+}
+
+func (t *Tracer) queryStats(query string) *queryStats {
+	t.mu.RLock()
+	stats := t.queries[query]
+	t.mu.RUnlock()
+
+	if stats != nil {
+		return stats
+	}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if t.queries == nil {
+		t.queries = make(map[string]*queryStats)
+	}
+	stats = t.queries[query]
+	if stats == nil {
+		stats = &queryStats{query: query}
+		t.queries[query] = stats
+	}
+	return stats
+}
+
+func (t *Tracer) collect() (rows []queryStats) {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	for query, s := range t.queries {
+		row := queryStats{
+			query:    query,
+			count:    atomic.LoadInt64(&s.count),
+			errors:   atomic.LoadInt64(&s.errors),
+			duration: atomic.LoadInt64(&s.duration),
+		}
+		rows = append(rows, row)
+	}
+	return rows
+}
+
+func (t *Tracer) Query(prepCtx context.Context, id sqlite.TraceConnID, query string, duration time.Duration, err error) {
+	stats := t.queryStats(query)
+
+	atomic.AddInt64(&stats.count, 1)
+	atomic.AddInt64(&stats.duration, int64(duration))
+	if err != nil {
+		atomic.AddInt64(&stats.errors, 1)
+	}
+}
```

Comment:
so this part runs on each query? a sync.Map for `t.queries` is probably a reasonable use of sync.Map considering that once it's seen each unique query it's never written to

Reply:
Note that only the first time a query is seen is the write mutex held. After that, only the read lock is needed. Given that means ~never holding the write lock in steady state I think this works out more efficient than a sync.Map.

Reply:
A sync.Map fast path is an atomic load. This PR's fast path is two atomic adds.

A sync.Map would be faster.

Reply:
Added a TODO, as I'm trying to clean my current open PRs, will come back to it.